internal/provider/connection_string_data_source_test.go

```diff
@@ -0,0 +1,145 @@
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package provider
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"testing"
+
+	"github.com/cockroachdb/cockroach-cloud-sdk-go/pkg/client"
+	mock_client "github.com/cockroachdb/terraform-provider-cockroach/mock"
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+// TestAccConnectionStringDataSource attempts to retrieve a cluster's
+// connection string. It will be skipped if TF_ACC isn't set.
+func TestAccConnectionStringDataSource(t *testing.T) {
+	t.Parallel()
+	clusterName := fmt.Sprintf("tftest-connect-%s", GenerateRandomString(4))
+
+	testConnectionStringDataSource(t, clusterName, false)
+}
+
+func TestIntegrationConnectionStringDataSource(t *testing.T) {
+	clusterName := fmt.Sprintf("tftest-connect-%s", GenerateRandomString(4))
+	clusterId := uuid.Nil.String()
+	if os.Getenv(CockroachAPIKey) == "" {
+		os.Setenv(CockroachAPIKey, "fake")
+	}
+
+	ctrl := gomock.NewController(t)
+	s := mock_client.NewMockService(ctrl)
+	defer HookGlobal(&NewService, func(c *client.Client) client.Service {
+		return s
+	})()
+
+	cluster := &client.Cluster{
+		Id:            clusterId,
+		Name:          clusterName,
+		CloudProvider: "GCP",
+		State:         "CREATED",
+		Config: client.ClusterConfig{
+			Serverless: &client.ServerlessClusterConfig{},
+		},
+		Regions: []client.Region{
+			{
+				Name: "us-central1",
+			},
+		},
+	}
+
+	db, sqlUser, os := "testdb", "test", "MAC"
+	connectStringOptions := &client.GetConnectionStringOptions{
+		Database: &db,
+		SqlUser:  &sqlUser,
+		Os:       &os,
+	}
+	fakeString := "postgresql://test@fake.cockroachlabs.cloud:26257/testdb"
+	connectionStringResponse := &client.GetConnectionStringResponse{
+		ConnectionString: &fakeString,
+		Params:           &map[string]string{},
+	}
+
+	httpOkResponse := &http.Response{Status: http.StatusText(http.StatusOK)}
+
+	s.EXPECT().CreateCluster(gomock.Any(), gomock.Any()).
+		Return(cluster, nil, nil)
+	s.EXPECT().GetCluster(gomock.Any(), clusterId).
+		Return(cluster, httpOkResponse, nil).Times(2)
+	s.EXPECT().GetConnectionString(gomock.Any(), clusterId, connectStringOptions).
+		Return(connectionStringResponse, httpOkResponse, nil).Times(4)
+	s.EXPECT().DeleteCluster(gomock.Any(), clusterId)
+
+	testConnectionStringDataSource(t, clusterName, true)
+}
+
+func testConnectionStringDataSource(t *testing.T, clusterName string, useMock bool) {
+	connectStrDataSourceName := "data.cockroach_connection_string.test"
+	sqlPassword := "hunter2"
```

Comment:
erm, all I see is `sqlPassword := "*******"`. Is github smart enough to know that `hunter2` is your password and show me `*******` instead?

Reply:
Yeah, I was a little worried after I realized that I used my actual 1Password master password here, but then I remembered that GitHub's got my back